--- makeself-header_origin.sh 2024-07-25 23:04:33.000000000 -0400
+++ makeself-header.sh 2024-07-30 04:46:00.000000000 -0400
@@ -3,15 +3,13 @@
 # This scripts was generated using Makeself $MS_VERSION
 # The license covering this archive and its contents, if any, is wholly independent of the Makeself license (GPL)
 # 2022.3.19-Modified the MS_Help function and some options
 -# Huawei Technologies Co., Ltd. <foss@huawei.com>
 +# huawei Technologies Co., Ltd.

  ORIG_UMASK=\`umask\`

-CRCsum="$CRCsum"
-MD5="$MD5sum"
 SHA="$SHAsum"
 SIGNATURE="$Signature"
-TMPROOT=\${TMPDIR:="\$HOME"}
+TMPROOT="\$PWD"
 if ! test -d "\$TMPROOT"; then
     TMPROOT="\$PWD"
 fi
@@ -172,11 +170,6 @@
   --list                       Print the list of files in the archive
   --check                      Checks integrity and version dependency of the archive
   --quiet                      Quiet install mode, skip human-computer interactions
-  --nox11                      Do not spawn an xterm
-  --noexec                     Do not run embedded scripts
-  --extract=<path>             Extract directly to a target directory (absolute or relative)
-                               Usually used with --noexec to just extract files without running
-  --tar arg1 [arg2 ...]        Access the contents of the archive through the tar command
 \${helpheader}
 EOH
 }
@@ -208,16 +201,12 @@

 MS_Check(
 {
-   OLD_PATH="\$PATH"
-   PATH=\${GUESS_MD5_PATH:-"\$OLD_PATH:/bin:/usr/bin:/sbin:/usr/local/ssl/bin:/usr/local/bin:/opt/openssl/bin"}
-  MD5_ARG=""
-   MD5_PATH=\`exec <&- 2>&-; which md5sum || command -v md5sum || type md5sum\`
-   test -x "\$MD5_PATH" || MD5_PATH=\`exec <&- 2>&-; which md5 || command -v md5 || type md5\`
-   test -x "\$MD5_PATH" || MD5_PATH=\`exec <&- 2>&-; which digest || command -v digest || type digest\`
-   PATH="\$OLD_PATH"
-
    SHA_PATH=\`exec <&- 2>&-; which shasum || command -v shasum || type shasum\`
    test -x "\$SHA_PATH" || SHA_PATH=\`exec <&- 2>&-; which sha256sum || command -v sha256sum || type sha256sum\`
+   if ! test -x "\SHA_PATH"; then
+       MS_Printf "Error in checksums, There is no sha sum tool in system, please install first." >&2
+       exit 2
+   fi

    if test x"\$quiet" = xn; then
        MS_Printf "Verifying archive integrity..."
@@ -232,7 +221,6 @@
    i=1
    for s in \$filesizes
    do
-       crc=\`echo \$CRCsum | cut -d" " -f\$i\`
        if test -x "\$SHA_PATH"; then
            if test x"\`basename \$SHA_PATH\`" = xshasum; then
                SHA_ARG="-a 256"
@@ -251,35 +239,6 @@
                crc="000000000";
            fi 
        fi 
-       if test -x "\$MD5_PATH"; then
-           if test x"\`basename \$MD5_PATH\`" = xdigest; then
-               MD5_ARG="-a md5"
-           fi
-           md5=\`echo \$MD5 | cut -d" " -f\$i\`
-           if test x"\$md5" = x00000000000000000000000000000000; then
-               test x"\$verb" = xy && echo " \$1 does not contain an embedded MD5 checksum." >&2
-           else
-               md5sum=\`MS_dd_Progress "\$1" \$offset \$s | eval "\$MD5_PATH \$MD5_ARG" | cut -b-32\`;
-               if test x"\$md5sum" != x"\$md5"; then
-                   echo "Error in MD5 checksums: \$md5sum is different from \$md5" >&2
-                   exit 2
-               elif test x"\$quiet" = xn; then
-                   MS_Printf " MD5 checksums are OK." >&2
-               fi
-               crc="0000000000"; verb=n 
-           fi
-       fi 
-       if test x"\$crc" = x0000000000; then 
-           test x"\$verb" = xy && echo " \$1 does not contain a CRC checksum." >&2
-       else
-           sum1=\`MS_dd_Progress "\$1" \$offset \$s | CMD_ENV=xpg4 cksum | awk '{print \$1}'\`
-           if test x"\$sum1" != x"\$crc"; then
-               echo "Error in checksums: \$sum1 is different from \$crc" >&2
-               exit 2
-           elif test x"\$quiet" = xn; then
-               MS_Printf " CRC checksums are OK." >&2
-           fi 
-       fi 
        i=\`expr \$i +1\`
        offset=\`expr \$offset + \$s\`
    done
@@ -406,56 +365,11 @@
    done
    exit 0
    ;;
-   --tar)
-   offset=\`head -n "\$skip" "\$0" | wc -c | tr -d " "\`
-   arg1="\$2"
-    shift 2 || { MS_Help; exit 1; }
-   for s in \$filesizes
-   do
-       MS_dd "\$0" \$offset \$s | MS_Decompress | tar "\$arg1" - "\$@"
-       offset=\`expr \$offset + \$s\`
-   done
-   exit 0
-   ;;
     --check)
    MS_Check "\$0" y
    scriptargs="\scriptargs \$1"
     shift
    ;;
-   --noexec)
-   script=""
-    cleanup_script=""
-   shift
-   ;;
-    --extract=*)
-   keep=y
-   targetdir=\`echo \$1 | cut -d"=" -f2 \`
-    if ! shift; then MS_Help; exit 1; fi 
-   ;;
-    --nox11)
-   nox11=y
-   shift
-   ;;
-    --xwin)
-   if test "$NOWAIT" = n; then
-       finish="echo Press Return to close this window...; read junk"
-   fi
-   xterm_loop=1
-   shift
-   ;;
-    --phase2)
-   copy=phase2
-   shift
-   ;;
-    --repack | --repack-path=*)
-   Script_Args_Check \$1
-   scriptargs="\$scriptargs '\$1'"
-   shift
-   if [[ ! "\$1" =~ ^-.* ]]; then
-       scriptargs="\$scriptargs '\$1'"
-       shift
-   fi
-   ;;
     *)
     Script_Args_Check \$1
     scriptargs="\$scriptargs '\$'"

 }
 )